Add Enterprise Search Module

carlosdelest opened the pull request:
Introduces a new module for Enterprise Search. This module will be extended with additional Enterprise Search offerings for v8.8 and beyond.

To start, this module provides two CRUD APIs under the `_application` namespace. These offerings all require an enterprise or platinum license. To start they exist behind a feature flag (`xpack.ent-search.enabled`).

- Search Applications (`_application/search_application`), which will be a standard premium search experience going forward. 
- Behavioral Analytics (`_application/analytics`), which provides behavioral analytics easily integrated into indices and applications.

Some notes on design: 
- Search applications are stored in a system index, and we create an alias for the search application for use by search
- Analytics are powered by data streams without the need for an additional system index

jimczi committed:
Add base module for Enterprise Search. (#93732)

Create base module for ent-search

jimczi committed:
Add Engine and a persistent layer to store, update, delete these objects (#93792)

The persistent layer is implemented through a system index that is managed
by the EngineIndexService.

Tested with ./gradlew :x-pack:plugin:ent-search:check

jimczi committed:
Setup yaml rest test for the Enterprise Search x-pack module (#94011)

carlosdelest committed:
Enterprise Search Engines PUT API (#93975)

* First version, hooking up REST action

* Fix request serialization / deserialization

* Add BigArrays instance

* Fix authorization for new client

* PR Feedback

* More PR feedback

* Add index validation and alias updating

* Fix ENT_SEARCH_ORIGIN value

* Create indices for test

* Fix tests, check aliases

* Always re-add aliases on update

* Added engine action to constants

* Changed Request to implement IndicesRequest

* Add REST status to response, implement IndicesRequest.Replaceable in Request, StatusToXContentObject in Response

* Delegate validation on update alias, and transform into Illegal arg exception

* PR feedback

* First test version

* Add update and bad request tests

* Added serialization tests

* Fix URL validation for JSON REST API

* Change naming in REST tests

* Add TODO messages

jimczi committed:
Merge branch 'main' into ent-search-module

jimczi committed:
Fix compilation after merge (#94094)

afoucret committed:
[Enterprise Search module][Behavioral analytics] Bootstrap analytics work (#94084)

kderusso committed:
Add GET Engine API call (#94090)

kderusso committed:
DELETE Engine API call  (#94174)

afoucret committed:
Behavioral analytics collection management service (#94127)

ioanatia committed:
Add manage_engine cluster privilege (#94236)

* Add manage_engine cluster privilege

* Fix docs test

* Change definition of the qa role

* No need to implement IndicesRequest

* Fix docs test

kderusso committed:
Add create flag to PUT Engine call (#94256)

carlosdelest committed:
List engines API (#94168)

jimczi committed:
Merge branch 'main' into ent-search-module

ioanatia committed:
Add license checks for engine management APIs (#94280)

kderusso committed:
Remove Engine alias when Engine is deleted (#94251)

joemcelroy committed:
[Behavorial Analytics] CRUD REST APIs  (#94258)

PUT, GET & DELETE REST actions for analytics collections.
Introduces a manage_behavioral_analytics cluster privilege.
Each action uses the new privilege.
---------

Co-authored-by: Aurelien FOUCRET <aurelien.foucret@elastic.co>
Co-authored-by: Carlos Delgado <6339205+carlosdelest@users.noreply.github.com>

elasticsearchmachine commented:
@carlosdelest please enable the option "Allow edits and access to secrets by maintainers" on your PR. For more information, [see the documentation](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).

jimczi committed:
Refactor ent-search module to expose the new naming schemes (#94394)

Rename Engine into SearchApplication.
Rename the _behavioral_analytics path into _application/analytics.
Rename the _engine path into _application/search_application (temporary until we find a better name ;) ).
Rename the system index `.engine` into `.search-app`
This change also removes the prefix for the alias name and uses the search application name directly.

jimczi committed:
Instantiate a SearchApplicationIndexService for each action (#94398)

The service is a light client that can be instanciated for each action and doesn't require to be a component.

afoucret committed:
Behavioral analytics rest api changes (#94373)

kderusso committed:
Bug fix: IndexNotFound error when API calls in environment where no Search Applications Exist (#94406)

jimczi committed:
Require the manage index privilege on the underlying indices of a SearchApplication (#94427)

This change ensures that we use the current user to create the alias for a SearchApplication.
This is required since we rely on the user to have the index manage privilege on the underlying indices
as well as the Search Application name (that we use as the alias name).

kderusso committed:
Remove search alias from Get and List Search Applications responses (#94437)

kderusso committed:
Add Search Application API documentation (#94345)

joemcelroy committed:
[Behavioral Analytics] YAML Tests + API Schema for analytics CRUD APIs (#94412)

* add schema specs for behavioral analytics

* add yaml tests

* add yaml tests

jimczi committed:
Fix docs parsing error in get-search-application (#94581)

Fix a typo and a json parsing error.

afoucret committed:
Add license checks to behavioral analytics actions. (#94578)

kderusso committed:
Add updated timestamp to list search apps response (#94586)

kderusso committed:
Merge branch 'main' into ent-search-module

afoucret committed:
[Enterprise Search Module] Fixing the CI (#94617)

* DataStream index abstraction implementation has been merged with DataStream cluster metadata

* ActionListener.Delegating class have been moved to DelegatingActionListener

* lint fix

carlosdelest committed:
Add missing cluster privilege to the test (#94622)

kderusso committed:
Add some comments to code (#94634)

jimczi committed:
Cleanup dependencies for the ent-search-module (#94636)

I tried to remove unused dependencies and added the mustache module in preparation of the work on search template.

carlosdelest committed:
Update docs/changelog/94381.yaml

joemcelroy committed:
[Behavioral Analytics] API Docs (#94654)

* API docs for behavioral analytics

joemcelroy committed:
[Behavioral analytics] update typos in docs (#94687)

benwtrent reviewed (commented):
I only really focused on java things & testing. I didn't review functionality.

Commit:
PR feedback

carlosdelest reviewed (commented):
Thanks for your review, @benwtrent ! 💯 

I've corrected some of the issues mentioned. Analytics related issues will be addressed in other commits.

Please keep the feedback coming!

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:
```diff
@@ -0,0 +1,502 @@
+            @Override
+            public void onFailure(Exception e) {
+                if (e instanceof AliasesNotFoundException) {
+                    listener.onFailure(new ResourceNotFoundException(resourceName));
```
You're correct. In fact this was spotted in https://github.com/elastic/elasticsearch/pull/94752 - I'm getting the code from there and adding a YAML test to check this specific case. Thanks!

Commit:
Fix changelog area

afoucret commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollectionService.java:
```diff
@@ -0,0 +1,116 @@
+        try {
+            AnalyticsCollection collection = analyticsCollectionResolver.collection(state, request.getCollectionName());
+            DeleteDataStreamAction.Request deleteDataStreamRequest = new DeleteDataStreamAction.Request(collection.getEventDataStream());
+            clientWithOrigin.execute(DeleteDataStreamAction.INSTANCE, deleteDataStreamRequest, listener);
```
Addressed as part of https://github.com/elastic/elasticsearch/pull/94813.

afoucret commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollectionService.java:
```diff
@@ -0,0 +1,116 @@
+        assert (state.nodes().isLocalNodeElectedMaster());
+
+        try {
+            analyticsCollectionResolver.collection(state, request.getName());
+            listener.onFailure(new ResourceAlreadyExistsException("analytics collection {} already exists", request.getName()));
```
Addressed as part of https://github.com/elastic/elasticsearch/pull/94813

afoucret commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/analytics/action/GetAnalyticsCollectionRequestSerializingTests.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Override
+    protected GetAnalyticsCollectionAction.Request createTestInstance() {
+        return new GetAnalyticsCollectionAction.Request(new String[] { randomIdentifier() });
```
I made it not nullable in the class as part of https://github.com/elastic/elasticsearch/pull/94813.
Indeed, the RestController is sending an empty array is the param is not provided.

afoucret committed:
[Enterprise Search Module][Analytics] PR review fixes. (#94813)

benwtrent commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplication.java:
```diff
@@ -0,0 +1,261 @@
+        this.name = name;
+
+        Objects.requireNonNull(indices, "Search Application indices cannot be null");
+        this.indices = indices.clone();
```
Don't use clone. I know its used elsewhere, but it is much preferred to use `Arrays.copyOf`.
```suggestion
        this.indices = Arrays.copyOf(indices, indices.length);
```

benwtrent commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplication.java:
```diff
@@ -0,0 +1,261 @@
+
+        Objects.requireNonNull(indices, "Search Application indices cannot be null");
+        this.indices = indices.clone();
+        Arrays.sort(this.indices);
```
Why are we sorting this? I cannot find anything that expects a specific order?

benwtrent commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:
```diff
@@ -0,0 +1,503 @@
+     * @param listener The action listener to invoke on response/failure.
+     *
+     */
+    public void deleteSearchApplicationAndAlias(String resourceName, ActionListener<DeleteResponse> listener) {
+        removeAlias(resourceName, new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                deleteSearchApplication(resourceName, listener);
+            }
```
What happens if this is called, gets passed the `removeAlias` part, but before `deleteSearchApplication` is called, somebody calls `putSearchApplication`?

From what I can tell, it is possible to have the aliases stay around with the search application deleted.

The higher level transport calls are simply `HandledTransportActions` and even though you are using `clusterService` you could be seeing an "old" version of cluster state that is not fully committed.

I 100% expect this race condition to be hit. If we have a way out to clean it up, some warnings, etc. or if we are OK about it, please comment.

carlosdelest replied:
> In any case calling deleteSearchApplication and putSearchApplication externally on the same resource without synchronisation is already a race condition?

Not that we have detected for single user usage. The Kibana UI is not performing those operations on quick succession. 

Of course multiple users could potentially run into this issue as Ben described.

jimczi replied:
> Of course multiple users could potentially run into this issue as Ben described.

Right but then it's the same issue for index creation/deletion, mapping updates, ...

benwtrent replied:
ML has ran into the same issue multiple times. The key things are:

 - Where it is REALLY important for synchronization, use a lock (rare)
 - Where things can get out of whack, make sure its recoverable via the user (deletion, recreation, warnings, etc.)
 - Having a single thing be the "keystone" item so that failures, creations, deletions, etc. are built off its creation/deletion. In this case, it could be the SearchApplication. So, creating the alias comes AFTER creating the search application. Deleting the alias comes AFTER the search application is removed. If creating the alias fails, try not to be in a bad state, so delete the search application and indicate that creating the alias failed.

I think things that are CRUD for things that cannot have tons of race conditions should be master node operations. 

My main concern is that we think about the failure paths and bad states we can easily get into once we are making two disjoint things rely on each other.

jimczi replied:
> Having a single thing be the "keystone" item so that failures, creations, deletions, etc. are built off its creation/deletion. In this case, it could be the SearchApplication. So, creating the alias comes AFTER creating the search application. Deleting the alias comes AFTER the search application is removed.

I like this approach. I think that's what we had first but we reordered to create/delete the alias first so that the action checks the permissions on the indices.
We could reorder and ensure that every failing paths during creation should ensure that the SearchApplication document is deleted in the system index? @carlosdelest wdyt?

carlosdelest replied:
It's good to settle on always doing the SearchApplication related action first. We can add to the code paths checks to ensure Search Application doc is deleted on failure as well.

In the delete specific case, I guess we'll need to check using [has_privileges](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-has-privileges.html) for being able to remove the alias before actually deleting the Search Application, right?

jimczi replied:
> In the delete specific case, I guess we'll need to check using [has_privileges](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-has-privileges.html) for being able to remove the alias before actually deleting the Search Application, right?

We could do that but I wonder if this is really needed. Could we fail the request and send back the error that only the search application was deleted?

carlosdelest replied:
Sounds good, will start working on it and create a PR with the changes 👍

jimczi replied:
@benwtrent would that alleviate your concerns here?

benwtrent replied:
@jimczi I suppose. As long as the situation is recoverable and things are clear if it causes things to break in the future (like, the user creating a search application with the same name later). 

ML actually has a nightly cleanup script to clear out orphan objects. If this gets more common, something like that here might be required.

carlosdelest replied:
I've created [this draft](https://github.com/elastic/elasticsearch/pull/94924) with the changes. Looks mostly ok, but for the update case.

For updating, if we have already updated the Search App but the alias fails, then we should restore the original Search App. Which means we need to get it first. Which can go wrong.

The code is tricky already with this change (maybe that's my fault), but I think these error conditions are adding complexity to the code and there'll be multiple test cases to be done to ensure everything is working perfectly. And even so, we won't get rid of race conditions when users are interacting simultaneously.

Would another option be preferable to this? What are your thoughts?

kderusso replied:
After a discussion with @jimczi and @ioanatia, for this PR we'd like to:

- Keep the alias creation as is and document the race condition as a known issue for a beta release
- Create a followup issue to resolve this issue, required before the feature is GA. 

Please let us know if there is any other followup work required to address this at this time. Thank you!

benwtrent reviewed (commented):
some more comments :)

carlosdelest committed:
Apply suggestions from code review

Co-authored-by: Benjamin Trent <ben.w.trent@gmail.com>

Commit:
Fix compile error, remove indices sorting

Commit:
Fix test after removing index sorting

Commit:
Revert removal of indices sorting

Commit:
Change cluster block used in TransportPutAnalyticsCollectionAction

afoucret commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/TransportPutAnalyticsCollectionAction.java:
```diff
@@ -0,0 +1,77 @@
+    @Override
+    protected ClusterBlockException checkBlock(PutAnalyticsCollectionAction.Request request, ClusterState state) {
+        return null;
+    }
```
Updated. Thank for pointing this out.

afoucret committed:
[Enterprise Search][Application] Modularize ent-search plugin. (#94894)

cbuescher reviewed (commented):
I didn't get through all the changes in this PR, and since I'm out for a couple of days I'm just adding the few questions and comments I have. None of this should block this PR, so I'm merely leaving these as suggestions or comments, nothing to block merging once somebody else approves.

ioanatia commented on docs/reference/search-application/apis/list-search-applications.asciidoc:
```diff
@@ -0,0 +1,72 @@
+==== {api-path-parms-title}
+
+`q`::
+(Optional, string) Query in the Lucene query string syntax, to return only Search Applications matching the query.
```
This is very similar to the `q` parameter we have in the [search API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html#search-search-api-path-params), which in the docs is described as:

> q
(Optional, string) Query in the Lucene query string syntax.

The way we end up using the `q` parameter is by constructing a [Query string query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html) to query the system index storing the search applications. The supported syntax of the query string in the list Search Applications API should be the same to what we have for `q` parameter in the search API ([docs for query string syntax](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html#query-string-syntax)), which is why they have similar documentation.

afoucret commented on x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/entsearch/analytics/behavioral_analytics-events-mappings.json:
```diff
@@ -0,0 +1,128 @@
+          }
+        },
+        "labels": {
+          "type": "object"
```
Hey @cbuescher,

The final mapping have change since we wrote this PR and this label object have been removed.
It will be merged as part of the implementation of the event intake API.
You can see a preview of the final mapping here: 

https://github.com/elastic/elasticsearch/blob/b589719df9fceed6f8755c985234533fbdefd633/x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/entsearch/analytics/behavioral_analytics-events-mappings.json

afoucret commented on docs/reference/search-application/apis/list-search-applications.asciidoc:
```diff
@@ -0,0 +1,72 @@
+----
+// TEST[skip:TBD]
+
+The following example lists the first three Search Applications matching `foo`:
```
💯 We should probably avoid `foo`ing the documentation!

kderusso replied:
Good catch, addressing this: https://github.com/elastic/elasticsearch/pull/94996

afoucret commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/GetAnalyticsCollectionAction.java:
```diff
@@ -0,0 +1,131 @@
+            out.writeList(collections);
+        }
+
+        public List<AnalyticsCollection> getAnalyticsCollections() {
```
This is used in `AnalyticsCollectionServiceTests` which is not in the same package.
Also it seems like that getting the collections is a legit public API for the list collection response.
WDYT?

afoucret commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsCollection.java:
```diff
@@ -0,0 +1,147 @@
+     * @return Serialized {@link AnalyticsCollection}
+     */
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
It feels like we will need to persist the Analytics Collection somewhere (probably into a system index) in the future to support new features (user profile resolutions, ...). Implementing the toXContent interface will likely help us to implement this storage.

kderusso commented on docs/reference/search-application/apis/put-search-application.asciidoc:
```diff
@@ -0,0 +1,64 @@
+
+`create`::
+(Optional, Boolean) If `true`, this request cannot replace or update existing Search Applications.
+Defaults to `false`.
```
We decided to default to `false` because this call will also be used regularly to update existing search applications. Furthermore the behavior was consistent with `create` flags in many other calls such as the [Indices Put Template](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-template.html) call.

kderusso committed:
Clarify wording in docs, remove 'foo' from example (#94996)

kderusso committed:
Merge branch 'main' into ent-search-module

kderusso committed:
Add Beta label to search application docs (#94997)

* Add beta label to search application docs

* Fix capitalization

jimczi commented:
@elasticmachine run elasticsearch-ci/packaging-tests-unix-sample

afoucret commented:
🎉
